src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -126,19 +135,35 @@ trait RegularSync {
         //we have same chain prefix
         if (parent.hash == headers.head.parentHash) {
 
-          val currentBranchTotalDifficulty = headersQueue.map(_.number)
-            .map(blockNumber => blockchain.getBlockByNumber(blockNumber))
-            .collect{
-              case Some(b) => b.header.difficulty
-              case None => BigInt(0)
-            }.sum
+          var oldBranch: Seq[Option[Block]] = Nil
```

Comment:
We can do it in a more functional way and get rid off those unnecessary Options:

```scala
          def getOldBlocks(headers: List[BlockHeader]): List[Block] = headers match {
            case h :: tail =>
              blockchain.getBlockByNumber(h.number).map(_ :: getOldBlocks(tail)).getOrElse(Nil)
            case Nil =>
              Nil
          }

          val oldBranch = getOldBlocks(headersQueue.toList)


          val currentBranchTotalDifficulty: BigInt = oldBranch.map(_.header.difficulty).sum

          val newBranchTotalDifficulty = headersQueue.map(_.difficulty).sum

          if (currentBranchTotalDifficulty < newBranchTotalDifficulty) {
            val transactionsToAdd = oldBranch.flatMap(_.body.transactionList)
            pendingTransactionsManager ! PendingTransactionsManager.AddTransactions(transactionsToAdd)
            val hashes = headersQueue.take(blockBodiesPerRequest).map(_.hash)
            waitingForActor = Some(context.actorOf(SyncBlockBodiesRequestHandler.props(peer, peerMessageBus, hashes)))
            //add first block from branch as ommer
            oldBranch.headOption.foreach { h => ommersPool ! AddOmmers(h.header) }
          } else {
            //add first block from branch as ommer
            headersQueue.headOption.foreach { h => ommersPool ! AddOmmers(h) }
            scheduleResume()
          }
```